RosNet/RosReader/RosBagReader.cs

```diff
@@ -2,67 +2,335 @@
+        byte[] latching = null;
+        while (reader.BaseStream.Position != endPos)
+        {
+            int fieldLen = reader.ReadInt32();
+            string fieldName = ReadName(reader, fieldLen);
+            byte[] fieldValue = reader.ReadBytes(fieldLen - fieldName.Length - 1);
+
+            switch (fieldName)
+            {
+                case "topic":
+                    originalTopic = fieldValue;
+                    break;
+                case "type":
+                    type = fieldValue;
+                    break;
+                case "md5sum":
+                    md5sum = fieldValue;
+                    break;
+                case "message_definition":
+                    messageDefinition = MessageDefinitionParser.ParseMessageDefinition(fieldValue);
+                    break;
+                case "callerid":
+                    callerID = fieldValue;
+                    break;
+                case "latching":
+                    latching = fieldValue;
+                    break;
+            }
+
+        }
+        connection.SetData(originalTopic, type, md5sum, messageDefinition, callerID, latching);
+    }
+
+    /// <summary>
+    /// Reads message data
+    /// </summary>
+    /// <returns>message data in bytes</returns>
+    private static byte[] ReadMessageData(BinaryReader reader)
+    {
+        var dataLength = reader.ReadInt32();
+        var data = reader.ReadBytes(dataLength);
+        return data;
+    }
+
+    /// <summary>
+    /// Reads chunk and adds all messages and message data to unParsedMessageHandler
+    /// </summary>
+    /// <returns>List of connections in chunk</returns>
+    private static List<Connection> ReadUnCompressedChunk(BinaryReader reader, UnParsedMessageHandler unParsedMessageHandler)
+    {
+        int dataLength = reader.ReadInt32();
+        List<Connection> connections = ReadChunk(reader, unParsedMessageHandler, dataLength);
+
+        return connections;
+
+    }
+
+    /// <summary>
+    /// Decompresses and reads chunk and adds all messages and message data to unParsedMessageHandler
+    /// </summary>
+    /// <returns>List of connections in chunk</returns>
+    private static List<Connection> ReadCompressedChunk(BinaryReader reader, UnParsedMessageHandler unParsedMessageHandler)
+    {
+        int compressedDataLength = reader.ReadInt32();
+        byte[] data = reader.ReadBytes(compressedDataLength);
+        byte[] unCompressedData = Array.Empty<byte>();
+        using MemoryStream source = new MemoryStream(data);
+        using MemoryStream target = new MemoryStream(compressedDataLength);
+        BZip2.Decompress(source, target, false);
+        unCompressedData = target.ToArray();
+
+        var connections = new List<Connection>();
+        using (BinaryReader tempReader = new BinaryReader(new MemoryStream(unCompressedData)))
+        {
+            connections = ReadChunk(tempReader, unParsedMessageHandler, unCompressedData.Length);
+        }
```

Comment:
```suggestion
        byte[] data = reader.ReadBytes(compressedDataLength);
        using var source = new MemoryStream(data);
        using var uncompressed = new MemoryStream(compressedDataLength);
        BZip2.Decompress(source, uncompressed, false);

        using var uncompressedReader = new BinaryReader(uncompressed);
        var connections = ReadChunk(uncompressedReader, unParsedMessageHandler, uncompressed.Length);
```
Why does this not work?

Reply:
As long as a Binary Reader can initialize with a memory stream, i dont see why it wouldnt

Reply:
When passing in length to ReadChunk, it expects an int, and the length of a MemoryStream is a long. I tried simply casting it to an int, and that cause some massive errors with reading beyond the length of the file

Reply:
I am guessing you would have to make the reader seek to the begining of the file again, e.g. `uncompressedReader.BaseStream.Seek(0, SeekOrigin.Begin)`

Reply:
Making it seek back to the beginning made it work, thanks